### PR TITLE
HPCC-29139 LDAP socket connections may need additional code to set ke…

### PR DESCRIPTION
**Pull request by RussWhitehead**
…epalive

Query config for keepAlive options, set them in LDAP when it is initialized

## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [x] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [x] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [x] The commit message title makes sense in a changelog, by itself.
  - [x] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [x] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [x] Thread-safety
  - [ ] Cloud-compatibility
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [x] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Smoketest:
- [ ] Send notifications about my Pull Request position in Smoketest queue.
- [ ] Test my draft Pull Request.

### CODE REVIEW

File: system/security/LdapSecurity/ldaputils.cpp [hunk 1]
```diff
@@ -68,6 +68,38 @@ LDAP* LdapUtils::LdapInit(const char* protocol, const char* host, int port, int
 #ifndef HAVE_TLS
         //throw MakeStringException(-1, "openldap client library libldap not compiled with TLS support");
 #endif
+
```
Comment (mckellyln) [inline system/security/LdapSecurity/ldaputils.cpp:71]: I think all this should be done both if ldaps or just ldap, so not in this if (stricmp(ldaps)) block but before or after.  And probably also ok if _WIN32.

**Review by mckellyln — CHANGES_REQUESTED** [review]

File: system/security/LdapSecurity/ldaputils.cpp [hunk 2]
```diff
@@ -105,7 +105,40 @@ LDAP* LdapUtils::LdapInit(const char* protocol, const char* host, int port, int
             return nullptr;
         }
 #endif
+
```
Comment (mckellyln) [inline system/security/LdapSecurity/ldaputils.cpp:108]: Do you have to call all the ldap_set_option()s before calling LDAP_INIT() ?

**Review by mckellyln — COMMENTED** [review]
Just have a question about where we need to set the options.

File: system/security/LdapSecurity/ldaputils.cpp [hunk 3]
```diff
@@ -105,7 +105,40 @@ LDAP* LdapUtils::LdapInit(const char* protocol, const char* host, int port, int
...
+    }
+
+    //Set TLS KeepAlive options
+    int kaTime;
+    int kaInterval;
+    int kaProbes;
+    if (queryKeepAlive(kaTime, kaInterval, kaProbes))//query OpenLDAP per connection tcp-keepalive settings
+    {
+        DBGLOG("queryKeepAlive returns time=%d, intvl=%d, probes=%d", kaTime, kaInterval, kaProbes);
```
Comment (mckellyln) [inline system/security/LdapSecurity/ldaputils.cpp:117]: skip DBGLOG ?  Perhaps change it to only log one line if any of the keepalive settings are made ?

**Review by mckellyln — CHANGES_REQUESTED** [review]
Just one change - to log keepalive settings a little differently, only log if any of them are not -1, so perhaps build up a string of what was changed and then log it as a single line.

File: system/security/LdapSecurity/ldaputils.cpp [hunk 4]
```diff
@@ -105,7 +105,41 @@ LDAP* LdapUtils::LdapInit(const char* protocol, const char* host, int port, int
...
+            rc = ldap_set_option(nullptr, LDAP_OPT_X_KEEPALIVE_INTERVAL, &kaInterval);
+            if(rc != LDAP_SUCCESS)
+                ERRLOG("LdapUtils::LdapInit : ldap_set_option(LDAP_OPT_X_KEEPALIVE_INTERVAL, %d) error - %s", kaInterval, ldap_err2string(rc));
+        }
+
+        if (kaProbes != -1)
+        {
+            rc = ldap_set_option(nullptr, LDAP_OPT_X_KEEPALIVE_PROBES, &kaProbes);
+            if(rc != LDAP_SUCCESS)
+                ERRLOG("LdapUtils::LdapInit : ldap_set_option(LDAP_OPT_X_KEEPALIVE_PROBES, %d) error - %s", kaProbes, ldap_err2string(rc));
```
Comment (mckellyln) [inline system/security/LdapSecurity/ldaputils.cpp:139]: if (kaTime != -1 || kaInterval != -1 || kaProbes != -1)
    {
        StringBuffer kaLogStr;
        if (kaTime != -1)
            kaLogStr.appendf(" time: %d", kaTime);
        if (kaInterval != -1)
            kaLogStr.appendf(" interval: %d", kaInterval);
        if (kaProbes != -1)
            kaLogStr.appendf(" probes: %d", kaProbes);
        DBGLOG("LDAP tcp keepalive%s", StringBuffer.str());
    }

File: system/security/LdapSecurity/ldaputils.cpp [hunk 5]
```diff
@@ -105,7 +105,41 @@ LDAP* LdapUtils::LdapInit(const char* protocol, const char* host, int port, int
...
+    }
+
+    //Set TLS KeepAlive options
+    int kaTime;
+    int kaInterval;
+    int kaProbes;
+    if (queryKeepAlive(kaTime, kaInterval, kaProbes))//query OpenLDAP per connection tcp-keepalive settings
+    {
+        if (kaTime != -1 || kaInterval != -1 || kaProbes != -1)
+            DBGLOG("queryKeepAlive returns time=%d, intvl=%d, probes=%d", kaTime, kaInterval, kaProbes);
```
Comment (mckellyln) [inline system/security/LdapSecurity/ldaputils.cpp:118]: ok, but "queryKeepAlive" maybe should be "LDAP tcp keepalive settings:" or something.  I was thinking more like:
```
StringBuffer kaLogStr;

rc = ldap_set_option()
if (rc != LDAP_SUCCESS)
     ERRLOG()
else
     kaLogStr.appendf(" time: %d", kaTime);
```
then at the end of the block:
```
if (!kaLogStr.isEmpty())
    DBGLOG("LDAP tcp keepalive%s", kaLogStr.str());
```

**Review by mckellyln — APPROVED** [review]
Looks good.

**Review by ghalliday — APPROVED** [review]

**Pull request merged**